tracepoint/src/native.rs

```diff
@@ -512,102 +559,36 @@ impl TracepointState {
         return writev_result;
     }
 
-    /// Fills in `data[0]` with the event's write_index, event_header,
-    /// activity extension block (if an activity id is provided), and
-    /// metadata extension block header (if meta_len != 0), then sends
-    /// the event to the `user_events_data` file.
+    /// Generates an event with headers.
     ///
-    /// Requires:
-    /// - `data[0].is_empty()` since it will be used for the headers.
-    /// - related_id may only be present if activity_id is present.
-    /// - if activity_id.is_some() || meta_len != 0 then event_header.flags
-    ///   must equal DefaultWithExtension.
-    /// - If meta_len != 0 then `data[1]` starts with metadata extension
-    ///   block data.
-    pub fn write_eventheader(
-        &self,
-        event_header: &EventHeader,
-        activity_id: Option<&[u8; 16]>,
-        related_id: Option<&[u8; 16]>,
-        meta_len: u16,
-        data: &mut [EventDataDescriptor],
-    ) -> i32 {
+    /// Requires: `data[0].is_empty()` since it will be used for the event headers;
+    /// `headers.len() >= 4` since it will be used for `write_index`.
+    ///
+    /// Returns 0 for success, error code (e.g. EBADF) for error. The error code
+    /// is usually ignored in retail scenarios but may be helpful during development to
+    /// understand behavior or track down issues.
+    ///
+    /// If disabled or unregistered, this method does nothing and returnes EBADF.
+    /// Otherwise, sets `data[0] = headers` and `headers[0..4] = write_index`, then sends
+    /// `data[..]` to the `user_events_data` file.
+    ///
+    /// The event's payload is the concatenation of the remaining data blocks, if any
+    /// (i.e. `data[1..]`).
+    ///
+    /// The payload's layout should match the args specified in the call to `register`.
+    pub fn write_with_headers(&self, data: &mut [EventDataDescriptor], headers: &mut [u8]) -> i32 {
```

Comment:
Note: `write_with_headers` is a new method, added to allow implementation of `write_eventheader` as a free function.